src/Bicep.Core/Workspaces/SourceFileGroupingBuilder.cs

```diff
@@ -161,291 +161,165 @@ private ResultWithDiagnostic<ISourceFile> PopulateRecursive(Uri fileUri, Artifac
 
         private void PopulateRecursive(BicepSourceFile file, IFeatureProviderFactory featureProviderFactory, IConfigurationManager configurationManager, ImmutableHashSet<ISourceFile>? sourceFilesToRebuild)
         {
-            var featureProvider = featureProviderFactory.GetFeatureProvider(file.FileUri);
-            if (featureProvider.ExtensibilityEnabled)
-            {
-                PopulateProviderDeclarations(file, featureProviderFactory, configurationManager);
-            }
-
-            foreach (var restorable in file.ProgramSyntax.Children.OfType<IArtifactReferenceSyntax>().Where(x => x is not ProviderDeclarationSyntax))
+            var config = configurationManager.GetConfiguration(file.FileUri);
+            foreach (var restorable in GetArtifactReferences(file.ProgramSyntax))
             {
-                var (childArtifactReference, uriResult) = GetArtifactRestoreResult(file.FileUri, restorable);
-                if (!uriResultByBicepSourceFileByArtifactReferenceSyntax.TryGetValue(file, out var uriResultByArtifactReferenceSyntaxLookup))
+                if (restorable is ProviderDeclarationSyntax providerDeclaration)
                 {
-                    uriResultByArtifactReferenceSyntaxLookup = [];
-                    uriResultByBicepSourceFileByArtifactReferenceSyntax[file] = uriResultByArtifactReferenceSyntaxLookup;
+                    var isBuiltInProvider = providerDeclaration.Specification switch {
+                        LegacyProviderSpecification => true,
+                        ConfigurationManagedProviderSpecification configSpec => config.ProvidersConfig.IsSysOrBuiltIn(configSpec.NamespaceIdentifier),
+                        _ => false,
+                    };
+
+                    if (isBuiltInProvider)
+                    {
+                        // built-in provider - no restoration required
+                        continue;
+                    }
+
+                    artifactLookup[restorable] = GetArtifactRestoreResult(file, restorable);
+
+                    // recursion not needed for provider declarations
+                    continue;
                 }
-                uriResultByArtifactReferenceSyntaxLookup[restorable] = uriResult;
 
-                if (!uriResult.IsSuccess(out var artifactUri))
+                var resolutionInfo = GetArtifactRestoreResult(file, restorable);
```

Comment:
Its possible to evaluate `GetArtifactRestoreResult(file, restorable);` once above the for-loop and use its result inside the if-clause rather than calling the function twice

Reply:
Could you explain why this is beneficial?

Reply:
This is just my opinion, but I think there are 2 main benefits:
1. As a reader it makes it clear that the artifact restore result doesn't depend on the restorable type, and that the restore result is terminal
2. It reduces the chance that one will change but not the other in future refactorings

These may not be strong enough justification, can you share your thoughts on the drawbacks of my suggestion and the benefits of keeping it as is?